### PR TITLE
Change BulkMoveWithWriteBarrier to be GC suspension friendly

### LINKED ISSUE #27769: Change BulkMoveWithWriteBarrier to be GC suspension friendly
PR https://github.com/dotnet/coreclr/pull/27758 reverted https://github.com/dotnet/coreclr/commit/5e1ef698774c433f70795e194e2554ad6f5b7d6f because it caused a failure in a Sockets test (https://github.com/dotnet/corefx/pull/42443#issuecomment-551155882, https://github.com/dotnet/corefx/pull/42443#issuecomment-551311471).  We should investigate and revert the revert with the appropriate fix.

I extracted a repro that doesn't depend on sockets (it's possible it could be simplified further, didn't spend more time on it):
```C#
    [Theory]
    [InlineData(1024)]
    public unsafe void Test(int copyLength)
    {
        var buffer1 = new byte[2400];
        for (int i = 0; i < buffer1.Length; i++) buffer1[i] = (byte)(i % 256);

        var buffer2 = new byte[buffer1.Length];
        var segments = new List<ArraySegment<byte>>();
        for (int i = 0; i < buffer2.Length; i++) segments.Add(new ArraySegment<byte>(buffer2, i, 1));

        var buffer1Handles = new List<GCHandle>();
        var buffer2Handles = new List<GCHandle>();
        for (int i = 0; i < buffer1.Length; i++) buffer1Handles.Add(GCHandle.Alloc(buffer1, GCHandleType.Pinned));
        for (int i = 0; i < buffer2.Length; i++) buffer2Handles.Add(GCHandle.Alloc(buffer2, GCHandleType.Pinned));

        int count = 0;
        while (segments.Count > 0)
        {
            int i;
            for (i = 0; i < copyLength && count < buffer1.Length; i++)
            {
                *(byte*)Marshal.UnsafeAddrOfPinnedArrayElement(buffer2, segments[i].Offset) = buffer1[count++];
            }
            segments.RemoveRange(0, i);
        }

        foreach (GCHandle handle in buffer1Handles) handle.Free();
        foreach (GCHandle handle in buffer2Handles) handle.Free();

        Assert.Equal(string.Join(",", buffer1), string.Join(",", buffer2));
    }
```
results in:
```
MyTests.Test(copyLength: 1024) [FAIL]
        Assert.Equal() Failure
                                         ↓ (pos 3658)
        Expected: ···1,252,253,254,255,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17···
        Actual:   ···1,252,253,254,255,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0···
                                         ↑ (pos 3658)
```
Presumably the changes from the PR come into play because of that RemoveRange call.

cc: @jkotas

### PR DESCRIPTION
**Pull request by jkotas**
Micro-benchmark results:

```
// Mean pause time: 
// Baseline: 32.9ms
// This change: 21ms
static void Main()
{
    new Thread(() =>
    {
        var src = new object[10_000_000]; 
        var dst = new object[10_000_000]; 
        for (;;) Array.Copy(src, dst, src.Length);
    }).Start();
    for (;;) { GC.Collect(); Thread.Sleep(1); }
}
```

```
// Mean pause time: 
// Baseline: 33.1ms
// This change: 0.8ms
static void Main()
{
    new Thread(() =>
    {
        var a = new byte[100_000_000];
        for (;;) a.Clone();
    }).Start();
    for (;;) { GC.Collect(); Thread.Sleep(1); }
}
```

Fixes #27769

### CODE REVIEW

**Comment by danmosemsft** [comment]
Do we have missing coverage in more specific corefx tests than networking? eg Array.Copy?

**Comment by jkotas** [comment]
The problem would be only hit for a very specific input: the size of the block must be more than 16kB, the array being copied must contain GC references, the destination must be overlap with source in specific way.  

I have changed the threshold for debug builds to be 1kB instead of 16kB that makes it much more likely for bugs like this to be caught. The original bug would by caught by more tests in CoreFX with the debug-build threshold.

File: src/System.Private.CoreLib/src/System/Buffer.CoreCLR.cs [hunk 1]
```diff
@@ -37,8 +39,57 @@ internal static unsafe void _ZeroMemory(ref byte b, nuint byteLength)
...
+        // Non-inlinable wrapper around the loop for copying large blocks in chunks
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void _BulkMoveWithWriteBarrier(ref byte destination, ref byte source, nuint byteCount)
+        {
+            Debug.Assert(byteCount > BulkMoveWithWriteBarrierChunk);
+
+            if (Unsafe.AreSame(ref source, ref destination))
+                return;
+
+            if ((nuint)(nint)Unsafe.ByteOffset(ref source, ref destination) >= byteCount)
```
Comment (VSadov) [inline src/System.Private.CoreLib/src/System/Buffer.CoreCLR.cs]: Indeed - `ByteOffset(a, b)`  is  `b - a`

Comment (VSadov) [reply src/System.Private.CoreLib/src/System/Buffer.CoreCLR.cs]: Easy to miss when reading late in the day. 
Maybe add a comment that this is folded    ` (destination - source)  >= byteCount || (destination - source) < 0 ` - just for the next one who reads this.

**Review by stephentoub — APPROVED** [review]

**Review by VSadov — APPROVED** [review]

**Comment by VSadov** [comment]
@stephentoub - about the test. Is that normal for sockets or the test was stressing some limits?

**Comment by stephentoub** [comment]
>  about the test. Is that normal for sockets or the test was stressing some limits?

The test was written to validate behavior when more segments are passed into a Receive/Send call than IOV_MAX, the maximum the number the platform can handle.  In that case we need to split the segments into multiple calls.  The test is validating that all data is sent/received appropriately when we need to do such handling.

I definitely don't expect this number of segments to be used in any real workload.  Generally the number is quite small, not 2400 :)

**Pull request merged**